### ISSUE TITLE
API server should return 202 if the operation is asynchronous

### ISSUE THREAD

**Issue by caesarxuchao** [kubernetes/kubernetes#33196]
Code 202 means "The request has been accepted for processing, but the processing has not been completed".

In k8s, the DELETE operation is asynchronous in two cases: 1. a graceful deletion of pod. 2. the object has pending finalizers. In both cases the object remains in the key-value store after the API server responds. The API server should return 202 in such cases. I haven't evaluated if such a change is backward compatible.

@mml @lavalamp @kubernetes/sig-api-machinery

**Comment by smarterclayton** [comment]
Yeah, I suspect this will break some error paths for some clients.  That said, those are probably bad clients - not sure how to get an accurate summary across the community.

**Comment by caesarxuchao** [comment]
@zreigz will be interested working on this? To work on this one, you'll need to know the concept of finalizers and pod graceful termination.

**Comment by zreigz** [comment]
LGTM If you have some docs or you can show me source code where the problem exists it would be helpful I will start investigation and let you know if I find some problems.

**Comment by zreigz** [comment]
I found some documentation and PR regarding those subjects. After reading this I will prepare POC for review.

**Comment by zreigz** [comment]
I've prepared some proposition: https://github.com/zreigz/kubernetes/commit/c650a6ac62e93e7dbadb87c0abd00b07deb860f1 please take a look. I am not sure about other places for modifications.

Another thing. I was testing it using the following commands:
```
kubectl --v=8 delete pod lukasz-3887938875-gn6y7 --grace-period=60
```
and the code for DELETE operation was 202 but when I use delete `deployment` then the value of `*options.GracePeriodSeconds` for each resource equals 0. The flag `--grace-period=60` is not taken into consideration. Is it OK? For me it looks like a bug.

**Comment by smarterclayton** [comment]
I'm concerned about the backwards compatible aspects of this - while it's the right thing to do, we need to be very sure we don't break working integrations.

**Comment by smarterclayton** [comment]
GracePeriod is only supported on Pods - for all other resources it is
treated as zero always.

**Comment by zreigz** [comment]
Thanks for explanation. I was testing my implementation  https://github.com/zreigz/kubernetes/commit/c650a6ac62e93e7dbadb87c0abd00b07deb860f1 where API server returns 202 with kubectl and everything works fine. Do you have some specific tests in mind. I can also run e2e tests.

**Comment by smarterclayton** [comment]
The "test" is how many clients out there check for status code 200 for
"delete succeeded" and would break.  A survey of the major language clients
for code that checks 200 on delete (as opposed to checking >= 200 && <300)
would be a start.

**Comment by caesarxuchao** [comment]
How about sending out a survey on https://groups.google.com/forum/#!forum/kubernetes-dev?

**Comment by smarterclayton** [comment]
We should do a code inspection too.  I'll search Java and Ruby clients if
you can do JS and Python?

On Oct 28, 2016, at 2:00 AM, Lukasz Zajaczkowski <notifications@github.com>
wrote:

You can find post with survey here:
https://groups.google.com/forum/#!topic/kubernetes-dev/8QI8bFl7Cdc

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/33196#issuecomment-256842862>,
or mute the thread

**Comment by bgrant0607** [comment]
Cascading deletion is new and optional, so it seems like we should be able to change that.

Pod graceful termination and namespace deletion would be the problematic cases.

**Comment by zreigz** [comment]
I will search JS and Python clients

**Comment by zreigz** [comment]
I've checked Python client from this list: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/client-libraries.md
It works fine when returned code is 202. Now I am going to check JS clients

**Comment by zreigz** [comment]
One more thing Python client doesn't support any flag (for example `--grace-period`) for delete operation. The same is for JS clients

**Comment by smarterclayton** [comment]
safe: abonas/kubeclient uses https://github.com/rest-client/rest-client
which handles 200-207 as success, but it only parses DELETE body on
exception, so 202 won't impact it either way
safe: https://github.com/doriordan/skuber looks like it handles <300 as
success and doesn't try to parse the body

unsafe:
https://github.com/fabric8io/kubernetes-client/blob/6e6690c18254b8a72fc9b202e214e2f9e33e1daf/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/OperationSupport.java
depends on the status code being 200 to know that delete is successful (so
we'd break this client)
unsafe:
https://bitbucket.org/amdatulabs/amdatu-kubernetes/src/629e81f2258cccb02231441ea3708e0d92c393f5/org.amdatu.kubernetes/src/org/amdatu/kubernetes/rest/KubernetesRestClient.java?at=master&fileviewer=file-view-default#KubernetesRestClient.java-360
checks 200 only

We'd need to contact these libraries and get them to relax their changes.
I'd be ok with contacting them informing them we will change in 1.6, but I
think given we have examples of broken clients I'd prefer not to break
people in 1.5 until we are sure they have updated.

**Comment by nikhiljindal** [comment]
@zreigz @smarterclayton Did we contact the owners of those clients?

I am interested in this for federation where cascading deletion (deleting the resource from underlying clusters) is asynchronous. federation apiserver should return 202 in those cases. Since kube-apiserver and federation-apiserver share common code, it will be easier to change both together.

**Comment by nikhiljindal** [comment]
Verified that those clients still have the problem.
Filed https://github.com/fabric8io/kubernetes-client/issues/647 for fabric8io and https://amdatu.atlassian.net/browse/AKC-2 for amdatu.

**Comment by jimmidyson** [comment]
As one of the authors of the fabric8 client, one of the bad ones 😜, just to confirm that we should really be checking status codes off all operations are in success range of >=200 && <300 ? Simple change we can make today & get a release out.

**Comment by smarterclayton** [comment]
Honestly this has backwards compatibility implications - we can't force
every client in the world to update by changing response codes without a
really good reason.  This breaks backwards compatibility, even if it is
more accurate.

**Comment by jimmidyson** [comment]
Totally agree, but I just wanted to be safe in case of changes in future.

**Comment by smarterclayton** [comment]
I'm not sure what boundary would be safe to change this on except an API
version boundary.

**Comment by nikhiljindal** [comment]
> As one of the authors of the fabric8 client, one of the bad ones ￼, just to confirm that we should really be checking status codes of all operations are in success range of >=200 && <300 ? Simple change we can make today & get a release out. /cc @iocanel

@jimmidyson Yes checking status code for all operations in success range of >=200 && <300 lgtm.

**Comment by slintes** [comment]
hey all, thanks for notifying us about this change, I updated the Amdatu Kubernetes client.
https://bitbucket.org/amdatulabs/amdatu-kubernetes/

**Comment by nikhiljindal** [comment]
Great thanks @slintes both the known bad clients have now been fixed!
That was fast :)

**Comment by smarterclayton** [comment]
We still can't change this - previous statements about backwards
compatibility holds.  Two examples that we know of means many more might be
broken that we don't.  I'm going to hold the line on compatibility here -
if there's a non-crazy boundary for a change, I'd support it.

**Comment by nikhiljindal** [comment]
> We still can't change this - previous statements about backwards
> compatibility holds.  Two examples that we know of means many more might be
> broken that we don't.

Agreed.



From @bgrant0607 :
> Cascading deletion is new and optional, so it seems like we should be able to change that.

How about if we detect that asynchronous deletion is happening because of orphanDependents and set the status to 202 only then?

**Comment by smarterclayton** [comment]
If orphanDependencies: false is opt in and we don't think anyone is using
it yet i think that's much safer.  That's also something we can handle in
the generic code (returning a metav1.Status{Code: Accepted} is trivial).
I'd be ok with that.

**Comment by nikhiljindal** [comment]
Great!

I am also planning to update the Delete() method in registry to return isDeleted boolean (similar to isCreated boolean returned by Update() that we use to return 201).
I can then return 202 if orphanDependents was false and registry returned isDeleted = false.
Does that sound good?
This is to still return 200 for resources that dont have cascading deletion even if user request contains OrphanDependents = false.

**Comment by nikhiljindal** [comment]
Sent https://github.com/kubernetes/kubernetes/pull/41165

**Comment by nikhiljindal** [comment]
#41165 is now merged (will be part of 1.6) and as discussed above we dont want to change the status code in other cases.
Closing the issue.
